### DISCUSSION TITLE
setIdx missing in Python wrapper

### DISCUSSION THREAD

**Discussion by filvuk** [rdkit/rdkit#6719] [Q&A] [answered]
I am trying to write atom map numbers provided from a SMILES tag into atom indices. For this, It would be most convenient to use the setIdx function. While this function seems to be documented in the C++ docs, I cannot find any reference to it in the Python docs, and I cannot call it with the Python wrapper. Is this intended or is there an available workaround?

**Comment by bp-kelley** [comment]
There is no setIdx on the atom, however to set atom map numbers, there is a "SetAtomMapNum"

```
>>> m = Chem.MolFromSmiles("CCC")
>>> for atom in m.GetAtoms():
...   atom.SetAtomMapNum(atom.GetIdx()+1) # 0 means no atom map
... 
>>> Chem.MolToSmiles(m)
'[CH3:1][CH2:2][CH3:3]'
```

**Comment by bp-kelley** [comment]
I should note that there is no setIdx in python, changing the idx of the atom would break your molecule and generally is done behind the scenes at molecule creation time.

**Reply by filvuk** [reply]
Thanks for the quick reply, first and foremost!

The python wrapper having no setIdx function is a bit unfortunate though in my case. Although it would be possible to define a mapping from atom map numbers to indices and performing all molecule operations via a translation layer, it seems like unnecessary overhead.
Is is alternatively possible to swap atom indices after molecule generation, or to regenerate a molecule with some specific instructions to atom indexing? Thanks in advance.

**Comment by DavidACosgrove** [comment, marked as answer]
You can reorder the atoms, as demonstrated here: https://github.com/rdkit/rdkit/discussions/4179
You can’t set an atom to have an arbitrary index. As @bp-kelley says, that will most likely break the molecule.

**Reply by filvuk** [reply]
Perfect, exactly what I've been looking for! Thanks.
